### ISSUE TITLE
Alpha-dropout for self-normalizing neural networks

### ISSUE THREAD

**Issue by anovstrup** [deeplearning4j/deeplearning4j#3739]
This [Self-Normalizing Neural Networks](https://arxiv.org/pdf/1706.02515.pdf) paper describes three features that the authors used to create fully-connected feed-forward neural networks that automatically control the mean/variance of activations in higher network layers: 
* a scaled exponential linear unit (SELU) activation function, already implemented by deeplearning4j/nd4j#1879.
* a new dropout technique ("alpha-dropout") that, during training, randomly sets inputs to the infimum of the SELU activation function (i.e., `-lambda * alpha`, where `lambda` and `alpha` are activation function parameters described in the paper).
* a new weight initialization scheme (similar to the MRSA/RELU scheme, but drawing from a normal distribution with variance `1/n` instead of `2/n`). This scheme is already implemented in DL4J as `WeightInit.XAVIER_FAN_IN`.

As far as I know, the alpha-dropout feature has not been implemented in DL4J.

I've taken a stab at implementing alpha-dropout as an InputPreProcessor, although I'm not sure how to ensure that the pre-processor is only applied to training passes:
```scala
class SeluInputPreProcessor(
    // the usual (confusingly named) dropout parameter
    // (i.e., the probability that a given input is *kept*)
    val dropout: Double) 
extends BaseInputPreProcessor {
  import ActivationSELU._

  private val ALPHA_PRIME = -SELU_LAMBDA * SELU_ALPHA
  private val a = math.pow(dropout + ALPHA_PRIME*ALPHA_PRIME*dropout*(1-dropout), -0.5)
  private val b = -a * (1-dropout) * ALPHA_PRIME

  override def getOutputType(inputType: InputType): InputType = inputType

  override def preProcess(input: INDArray, miniBatchSize: Int): INDArray = {
    val keep: INDArray = Nd4j.rand(input.shape).ltei(dropout)
    input.muli(keep)
    input.addi(Nd4j.getExecutioner.execAndReturn(new OneMinus(keep)).muli(ALPHA_PRIME))
    input.muli(a)
    input.addi(b)
    input
  }

  override def backprop(output: INDArray, miniBatchSize: Int): INDArray = output
}
``` 

I _think_ that this implementation of the alpha-dropout transformation is correct, but it should be double-checked against the paper and I'm not at all sure that's it's optimal.

**Comment by AlexDBlack** [comment]
Apologies for not getting to this sooner. We have 2 options here - though I'm inclined to implement them both.

(a) We'll change the preprocessor API to have a 'train vs. test' arg. This may come up in some other uses cases, and there's no real reason not to add it.

(b) We could implement this as built-in functionality for layers, rather than a dropout preprocessor... perhaps configured via a DropoutType enumeration, with values {STANDARD, ALPHA} initially, and maybe others (variational dropout?) in the future.

**Comment by anovstrup** [comment]
I agree that it's probably worthwhile to implement both. A minor note with (b) is that some users may prefer to set the _alpha_ and _lambda_ parameters in the SELU activation function (although the current implementation doesn't support this anyway). If those parameters do become configurable, then the implementation of alpha-dropout would need to be updated to employ the same parameters as the SELU layers.

**Comment by raver119** [comment]
side note: I don't see any significant blockers to make selu_alpha & selu_lambda configurable.

**Comment by anovstrup** [comment]
Oh, another note regarding the implementation above. As far as I know, it does not support setting the seed for the random number generator, which would be a problem for users that desire repeatability. I expect that implementing alpha-dropout as built-in layer functionality, as in (b), would address this issue (just as, I assume, the existing dropout implementation does).

**Comment by AlexDBlack** [comment]
Added here: https://github.com/deeplearning4j/deeplearning4j/pull/3985
